lib/llhttp/parser.js

```diff
@@ -0,0 +1,223 @@
+
+const kPtr = Symbol('kPrt');
+const kUrl = Symbol('kUrl');
+const kStatusMessage = Symbol('kStatusMessage');
+const kHeadersFields = Symbol('kHeadersFields');
+const kHeadersValues = Symbol('kHeadersValues');
+const kBody = Symbol('kBody');
+const kReset = Symbol('kReset');
+const kCheckErr = Symbol('kCheckErr');
+
+const cstr = (ptr, len) =>
+  Buffer.from(inst.exports.memory.buffer, ptr, len).toString();
+
+const wasm_on_message_begin = p => {
+  const i = instMap.get(p);
+  i[kReset]();
+  return i[kOnMessageBegin]();
+};
+
+const wasm_on_url = (p, at, length) => {
+  instMap.get(p)[kUrl] = cstr(at, length);
+  return 0;
+};
+
+const wasm_on_status = (p, at, length) => {
+  instMap.get(p)[kStatusMessage] = cstr(at, length);
+  return 0;
+};
+
+const wasm_on_header_field = (p, at, length) => {
+  const i= instMap.get(p)
+  i[kHeadersFields].push(cstr(at, length));
+  return 0;
+};
+
+const wasm_on_header_value = (p, at, length) => {
+  const i = instMap.get(p);
+  i[kHeadersValues].push(cstr(at, length));
+  return 0;
+};
+
+const wasm_on_headers_complete = p => {
+  const i = instMap.get(p);
+  const type = inst.exports.llhttp_get_type(p);
+  const versionMajor = inst.exports.llhttp_get_http_major(p);
+  const versionMinor = inst.exports.llhttp_get_http_minor(p);
+  const rawHeaders = [];
+  let method;
+  let url;
+  let statusCode;
+  let statusMessage;
+  const upgrade = inst.exports.llhttp_get_upgrade(p);
+  const shouldKeepAlive = inst.exports.llhttp_should_keep_alive(p);
+
+  for (let c = 0; c < i[kHeadersFields].length; c++) {
+    rawHeaders.push(i[kHeadersFields][c], i[kHeadersValues][c])
+  }
+  i[kOnHeaders](rawHeaders);
+
+  if (type === HTTPParser.REQUEST) {
+    method = constants.METHODS[inst.exports.llhttp_get_method(p)];
+    url = i[kUrl];
+  } else if (type === HTTPParser.RESPONSE) {
+    statusCode = inst.exports.llhttp_get_status_code(p);
+    statusMessage = i[kStatusMessage];
+  }
+  return i[kOnHeadersComplete](versionMajor, versionMinor, rawHeaders, method,
+url, statusCode, statusMessage, upgrade, shouldKeepAlive);
+};
+
+const wasm_on_body = (p, at, length) => {
+  const i = instMap.get(p);
+  const body = Buffer.from(inst.exports.memory.buffer, at, length);
+  return i[kOnBody](body);
+};
+
+const wasm_on_message_complete = (p) => {
+  return instMap.get(p)[kOnMessageComplete]();
+};
+
+const instMap = new Map();
+
+const inst = new WebAssembly.Instance(mod, {
+  env: {
+    wasm_on_message_begin,
+    wasm_on_url,
+    wasm_on_status,
+    wasm_on_header_field,
+    wasm_on_header_value,
+    wasm_on_headers_complete,
+    wasm_on_body,
+    wasm_on_message_complete,
+  },
+});
+
+inst.exports._initialize(); // wasi reactor
+
+class HTTPParser {
+  static REQUEST = REQUEST;
+  static RESPONSE = RESPONSE;
+  static kOnMessageBegin = kOnMessageBegin;
+  static kOnHeaders = kOnHeaders;
+  static kOnHeadersComplete = kOnHeadersComplete;
+  static kOnBody = kOnBody;
+  static kOnMessageComplete = kOnMessageComplete;
```

Comment:
Remember we need this to work on Node 12 as well.

Reply:
Ok 👍🏻 . Are you referring to something already in these changes that is not supported?

Reply:
I don't think static works?

Reply:
I'll check it out 👍🏻

Reply:
It seems supported on 12, not on 10, however. `standard` complains about it on all versions 😄 .